pkg/alertdog/alertdog.go

```diff
@@ -0,0 +1,165 @@
+package alertdog
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/PagerDuty/go-pagerduty"
+	"github.com/prometheus/alertmanager/template"
+
+	"github.com/errm/alertdog/pkg/alertmanager"
+)
+
+type Alertmanager interface {
+	Alert(alertmanager.Alert) error
+	Resolve(alertmanager.Alert) error
+}
+
+type Pagerduty interface {
+	ManageEvent(event pagerduty.V2Event) (*pagerduty.V2EventResponse, error)
+}
+
+type Alertdog struct {
+	AlertmanagerEndpoints []string `yaml:"alertmanager_endpoints"`
+	Expected              []*Prometheus
+	CheckInterval         time.Duration
+	Expiry                time.Duration
+	PagerDutyKey          string
+	PagerDutyRunbookURL   string
+
+	mu           sync.RWMutex
+	checkedIn    time.Time
+	alertmanager Alertmanager
+	pagerduty    Pagerduty
+}
+
+func (a *Alertdog) UnmarshalYAML(unmarshal func(interface{}) error) error {
```

Comment:
Where is this function used and what's its use? It seems like you're always expecting `unmarshall` to be a `func(*Alertdog)`, so why not have that instead and simplify this a bit?

Reply:
This is so we can set defaults for `CheckInterval` and `Expiry`

The type aliasing thing is to avoid us recursivly calling this function!

Reply:
It is not used on those functions though. :thinking:

Reply:
`CheckInterval` and `Expiry` are fields that we expect to configure from the yaml config file...

Here we are setting defaults incase you didn't set them.

Reply:
Maybe @mtpereira is missing https://pkg.go.dev/gopkg.in/yaml.v2#Unmarshaler ?

Reply:
Also: cute trick.

Reply:
> Also: cute trick.

it is somewhere between a horrible hack and an elegent solution ...